src/tests/ftest/util/general_utils.py

```diff
@@ -663,16 +668,34 @@ def stop_processes(hosts, pattern, verbose=True, timeout=60, added_filter=None):
     """
     result = {}
     log = getLogger()
-    log.info("Killing any processes on %s that match: %s", hosts, pattern)
+    if send_sigusr2 is True:
+        log.info("First dumping ULT stacks, then Killing any processes on %s "
+                 "that match: %s", hosts, pattern)
+    else:
+        log.info("Killing any processes on %s that match: %s", hosts, pattern)
 
     if added_filter:
-        ps_cmd = "/usr/bin/ps x | grep -E {} | grep -vE {}".format(
+        ps_cmd = "/usr/bin/ps xa | grep -E {} | grep -vE {}".format(
             pattern, added_filter)
     else:
         ps_cmd = "/usr/bin/pgrep --list-full {}".format(pattern)
 
     if hosts is not None:
-        commands = [
+        if send_sigusr2 is True:
+            commands_part1 = [
+                "rc=0",
+                "if " + ps_cmd,
+                "then rc=1",
+                "sudo pkill --signal USR2 {}".format(pattern),
+                # leave time for ABT info/stacks dump vs xstream/pool/ULT number
+		"sleep 20",
```

Comment:
Sleeps are bad.  A polling loop with a timeout for the outlier cases would be better as it does not add 20 seconds to processing for every single case, only the outlier ones.

Reply:
Well, the problem here is that there is no real mean to check if dump of ULT stacks is done/stuck, because it is run under direct control of ABT in a completely asynchronous way ... So the 20s is an approximation to leave enough time for it to complete.

Reply:
The result is the writing of files though, yes?  If so you could poll the size of the file ([`stat(1)`](https://linux.die.net/man/1/stat)) in a loop wait until it's not being written any more, and/or check whether it's open ([`fuser(1)`](https://linux.die.net/man/1/fuser)) by any process, etc.

Anyway, my caution here is just that @johannlombardi has identified sleeps in our test framework as potentially being the cause of general unnecessary slowness in our testing throughput.  I'm not sure how frequently this sleep will be hit in a test run, but just caution that while 20s doesn't seem like a lot on it's own, the cumulative effect can be costly if it's called many times.  Ultimately you can be the judge of whether that is a potential problem or not and decide if it needs addressing but just keep in mind how many times we can do repetitive operations in a test run.  I.e. we don't just start servers once for a whole test run but start and stop them for every single test (but not test variants, any more) in that test run.

Reply:
Well, the stat() loop is not easy to implement as the file filling appears to not occur linearly. And file is never closed until engine's exit. Last, the branch with the sleep should only be executed upon detection of tests hangs/timeouts situation and when engines are found still running.